Make our Dangerzone image reproducible

apyrgio opened the pull request:
This PR makes our container image reproducible, and enforces it with a CI job on every commit. To do so, we also change our base image from Alpine Linux to Debian Stable, and we use some prior art and tooling from the work that @AkihiroSuda has done on reproducible containers.

Fixes #1046
Fixes #1047 
Fixes #1048

almet commented on docs/developer/reproducibility.md:
````diff
@@ -0,0 +1,125 @@
+      --ignore-timestamps --ignore-image-name --verbose
+  ```
+
+### Updating the image
````
I feel we're missing a howto check reproducibility for old images. Maybe rephrasing a bit this section could help in this regard (as I believe it would be updating the version numbers / dates and rebuilding the image, then diffing)

apyrgio replied:
I already have a "Reproducing the image" section in the design doc, not sure if you've seen it. If it's not clear what the user should do, here's an update in the wording: aa710e8

almet replied:
Yep, it's better as it explains how to do it for older images, thanks!

almet commented on install/common/build-image.py:
```diff
@@ -64,111 +67,78 @@ def main():
+        # 3. If the contents of the Git repo are dirty, we will append a unique identifier
+        #    for this run, something like `0.8.0-31-g6bdaa7a-fdcb` or `0.8.0-fdcb`.
+        dirty_ident = secrets.token_hex(2)
+        tag = (
```
That could go in its own function to ease readability.

apyrgio replied:
Implemented in 92d8a4c

almet replied:
Thanks. I wonder if mentioning it's getting the tag name from git might help, since we might also have tags for the image.

almet replied:
(in the name of the function, I mean)

apyrgio replied:
Hm, doesn't hurt. Check 0f0fa49923ef42d01c96b0e4163292f2f580279d.

almet commented on install/linux/vendor-pymupdf.py:
```diff
@@ -28,7 +28,7 @@ def main():
     )
```
Reading this, I wonder if we should rename the `vendor-pymupdf` to something specifically targeting debian-based distributions.

almet replied:
After discussion, we've decided to go with `debian-vendor-pymupdf.py`

almet commented on Dockerfile:
```diff
@@ -1,102 +1,81 @@
+
+FROM debian:bookworm-${DEBIAN_IMAGE_DATE}-slim
+
+ARG GVISOR_ARCHIVE_DATE=20250106
```
Also, should we get these directly from the `Dockerfile.env` file instead of defining them twice?

apyrgio replied:
Very good point. So, I've created an issue for automating the bump of these values, since I wanted to pursue it in this PR, but I didn't have time (check #1050). As for the URLs, I've added them to `Dockerfile.env` as comments (see 96ab442873f32b650da642f487bb0702c30cb304, which also fixes a bug). As for the values in `Dockerfile`, they are actually enerated from the following command:

```
poetry run jinja2 Dockerfile.in Dockerfile.env > Dockerfile
```

So, they are practically defined once. Are you ok with that?

apyrgio replied:
I think it doesn't hurt committing it, and it's easier for viewers of the repo to see at a glance what are the contents of the image.

almet replied:
On the "cons" side, it means that all our changes to the containers will be duplicated in these two files each time we do a change. I would prefer otherwise but can live with it if you have strong feeling on this.

If we decide to go down that road, we might want to have a CI task ensuring that `Docker` is then the output of `Docker.in` processed so the two don't end up out of sync.

apyrgio replied:
> If we decide to go down that road, we might want to have a CI task ensuring that Docker is then the output of Docker.in processed so the two don't end up out of sync.

Yeap, I agree. As I mentioned in https://github.com/freedomofpress/dangerzone/pull/1049#discussion_r191475956, there's one already in place:

https://github.com/freedomofpress/dangerzone/blob/b8755f26ee56fdb76389049c96560e692a4fb4e2/.github/workflows/ci.yml#L490-L493

almet replied:
Okay, marking this is resolved then.

almet reviewed (commented):
Nice work on this!

I added a few comments here and there, and will probably come back to it once I've run it on my laptop. It's pretty neat!

apyrgio commented on .github/workflows/ci.yml:
```diff
@@ -472,3 +471,41 @@ jobs:
           xvfb-run -s '-ac' ./dev_scripts/env.py --distro ${{ matrix.distro }} --version ${{ matrix.version }} run --dev \
               bash -c 'cd dangerzone; poetry run make test'
+
+  reproduce-image:
```
Done in 10b8cd48afb5fc4215b90f96791bc91aafbea43f.

almet commented on .github/workflows/build.yml:
```diff
@@ -85,7 +85,7 @@ jobs:
         uses: actions/cache@v4
         with:
-          key: v3-${{ steps.date.outputs.date }}-${{ hashFiles('Dockerfile', 'dangerzone/conversion/common.py', 'dangerzone/conversion/doc_to_pixels.py', 'dangerzone/conversion/pixels_to_pdf.py', 'poetry.lock', 'gvisor_wrapper/entrypoint.py') }}
+          key: v3-${{ steps.date.outputs.date }}-${{ hashFiles('Dockerfile', 'dangerzone/conversion/*.py', 'dangerzone/container/*', 'install/common/build-image.py') }}
```
> since we install PyMuPDF from the Debian repos.

Okay, I think I get it now: we still install PyMuPDF (via poetry), but only for the user-facing application. Hence why we can remove the `poetry.lock` file from here. :+1:

apyrgio commented on Dockerfile:
```diff
@@ -1,102 +1,81 @@
+    conversion/common.py \
+    conversion/errors.py \
+    conversion/__init__.py \
+    /opt/dangerzone/dangerzone/conversion
```
I'm copying the `.py` files specifically for two reasons: a) to avoid picking up the `.pyc` files, and b) to make the list of copied files more explicit for reproducibility reasons. As for having a separate package, I think it will add a bit more complexity than strictly necessary for this invocation. I'd like to know if you have something particular in mind though.

apyrgio replied:
Implemented in 9daf301

apyrgio commented on dangerzone/container/repro-sources-list.sh:
```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
```
It has to be withing the `dangerzone/` dir, because that's the build context of the `podman build` command. It can't be outside of it, I'm afraid. Wanna rename the `container/` dir though? I don't like the name to be honest, but maybe you have something better.

almet replied:
Thanks for the explanation. I actually got confused by the `container` name here, as we have a `isolation_provider/container.py` file.

For the files that needs to be accessed by the container, we could put them together inside a folder where we could regroup `gvisor_wrapper/entrypoint.py` and `container/repro-sources-list.sh` (for now).

As for the name, what do you think about `container_scripts`? (note, we also have a `/container_utils.py` file.

almet replied:
Discussed this and decided on `container_helpers`

apyrgio replied:
Implemented in 02e63e5

apyrgio commented on dangerzone/conversion/doc_to_pixels.py:
```diff
@@ -129,6 +129,8 @@ async def convert(self) -> None:
             # At least .odt, .docx, .odg, .odp, .ods, and .pptx
             "application/zip": {
                 "type": "libreoffice",
+                # NOTE: Older `file` command cannot detect hwpx files properly.
```
It's file `5.45` I think, based on https://github.com/freedomofpress/dangerzone/pull/460#issuecomment-1654166465 and https://github.com/file/file/tags. Debian stable is currently in `5.44` (see https://packages.debian.org/bookworm/file).

almet replied:
Might be good to put that version in the comment, so we know when it can cause problems.

apyrgio replied:
Added in 2b71e615a8157cf3e8a71c56fd3a3626b70a7206

apyrgio commented on dev_scripts/reproduce.py:
```diff
@@ -0,0 +1,151 @@
+        args,
+        check=True,
+        stdout=subprocess.PIPE,
+        # stderr=subprocess.PIPE
```
Yeap, removed in b8755f26ee56fdb76389049c96560e692a4fb4e2.

apyrgio commented on docs/developer/reproducibility.md:
````diff
@@ -0,0 +1,125 @@
+  ```
+  ./diffoci diff podman://<new_image_tag> podman://<old_image_tag> \
+      --ignore-timestamps --ignore-image-name --verbose
+  ```
````
Right, I'm mixing two things here, mostly because I was in a hurry :grimacing: :
* Design document material, which tend to become stale since it reflects how we _will_ try to solve a problem. Design docs tend to age relatively fast, since they are not supposed to be updated (usually). A recent example is the gVisor docs.
* Build docs on how to check for reproducibility and bump image params. These instructions will be part of our release instructions, so we should keep them fresh for a while.

Ultimately, what you're proposing makes sense, if we want to keep just the fresh part of the docs in our repo. If we want to keep our design document as well, and expect it to age, we can move the commands to our `BUILD.md`. Let me know what you prefer.

almet replied:
After a discussion, we decided to only include general context and a how to check reproducibility. This is also tied to #903 where the documentation can be improved in general for DZ.

apyrgio replied:
Improved in acbc43371747b9e9917590039fa6c0fd3d04ccb5

apyrgio commented on dev_scripts/reproduce.py:
```diff
@@ -0,0 +1,151 @@
+
+    if not DIFFOCI_PATH.exists():
+        logger.info(f"Downloading diffoci helper from {DIFFOCI_URL}")
+        diffoci_download()
```
Implemented in c1f2548

almet replied:
As discussed, I think the current implementation lacks two different things:

1. it should probably be `if not diffoci_exists()` ;
2. the name of the function could be changed to something closer to the meaning of what it provides. like `diffoci_is_installed`.

apyrgio replied:
Fixed in 685cf431a337dd0da986f8b5e22602d657375388.

almet commented on dangerzone/container/repro-sources-list.sh:
```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+#
```
We also had a discussion about this file, and if it should be included in this repository or not (the other option would be to download it on the fly, which could be done by the `build-image.py` script.).

We didn't came up with a conclusion on this one just yet. It feels odd to have it included here, but might work on the short run.

apyrgio replied:
Hm, would a third party notice make things better? Check out 4595b3ad4a834cef307c06d1021069b19502522d

almet commented on dev_scripts/reproduce.py:
```diff
@@ -0,0 +1,166 @@
+
+logger = logging.getLogger(__name__)
+
+DIFFOCI_URL = "https://github.com/reproducible-containers/diffoci/releases/download/v0.1.5/diffoci-v0.1.5.linux-amd64"
```
It seems to me that `reproduce.py` requires `diffoci` but only checks out the `amd64` version. Should we also find a way to download it on `arm64` (e.g. on silicon macs)?

almet replied:
Yep, we can go with this for now and refine later!

almet commented on dev_scripts/reproduce.py:
```diff
@@ -0,0 +1,166 @@
+    parser.add_argument(
+        "--source",
+        required=True,
+        help="The source image name that you want to reproduce (in diffoci format)",
```
I found it useful to have the doc handy to run `reproduce.py`, especially for the arguments to pass to `--source`. Having an example defined here might help users of the script.

apyrgio replied:
Took a stab in 35802e78981b194e8b1e694aa152ae03510112a6

almet commented on dev_scripts/reproduce.py:
```diff
@@ -0,0 +1,166 @@
+def parse_args():
+    parser = argparse.ArgumentParser(
+        prog=sys.argv[0],
+        description="Dev script for verifying container image reproducibility",
```
We might want to point people to the documentation for more info.

almet replied:
This is ewww :star_struck:

almet reviewed (commented):
I've checked-out your branch and used it locally. I'm glad to report that it works well! :tada: 

I've added a few other questions / remarks, the main one being about ARM64 support for reproducibility.

almet commented:
(Also happy to report that the size of the image dropped from 577 MB to 445 MB with the changes in this branch)

almet reviewed (approved):
Thanks for all the changes, I believe we're good to merge as soon as the CI is green !

apyrgio commented:
I've updated this PR with a way to separate the rootfs of the inner and outer container image, but reuse the same `/usr` dir, which takes the main bulk of the space. Check out:
* https://github.com/freedomofpress/dangerzone/issues/1048#issuecomment-2610175596
* https://github.com/freedomofpress/dangerzone/pull/1049/commits/414efb629feab21fc42886e64fce804a6d9effe6

I've also rebased it on top of the `platformdirs` branch, so that the tests can pass. I believe we can merge the symlink approach for now, with the caveat that it's too "magic". I'd prefer something more straight-forward, but in this case it's not that simple.

Let me know your thoughts, thanks!

apyrgio commented:
Sigh, I just realized something important while scanning the final image with Grype; Grype does not detect vulnerabilities for the inner container image. It could do that for the Alpine one, but it seems it doesn't do it for the Debian one. This could also be due to the fact that we use scratch images (see https://github.com/anchore/grype/issues/383), or because we haven't copied the `/var` layer into the inner container image, which has some `apt` info.

So, it seems that I need to dig more into it :-/
